alphagov/paas-bootstrap
# [#131876783] Migration from paas-cf to paas-bootstrap

[#131876783 Migrate paas-cf to paas-bootstrap](https://www.pivotaltracker.com/story/show/131876783)

**NB: Please review but do not merge over holiday period**

What?
-----

We want to use paas-bootstrap project to create the initial VPC, bosh and concourse deployer, and remove the required logic from paas-cf.

For that, apart of migrate the logic, we must also implement a procedure to migrate the existing environments.

In this PR we :
 * Move the basic configuration to get VPC, bosh and concourse running from paas-cf.
 * Implement logic to migrate an existing environment:
   * Copy blobstores across, as they get renamed.
   * Recreate the existing bosh VMs to allow change the blobstore.
   * Reattach the old volume from concourse into the new one.

All the migration steps are idempotent and the commits can be reverted later after the migration is done in all the environments.

Note: There are several features and small useful scripts that were not migrated yet. The change is big enough and will be done in follow up PRs.

Dependencies
-----------------

This PR must be reviewed, tested and applied in conjuction with https://github.com/alphagov/paas-cf/pull/688

How to review, test and apply it?
---------

Do a code review to check that makes sense.

To test it: 

 1. Deploy a CF environment using the master branch of  https://github.com/alphagov/paas-cf
 2. Apply the procedure described below.

Migration procedure
------------------------

This branch includes all the required logic almost do all the migration from a paas-cf deployment. The steps are in the commits prefixed with `migrate`.

The full migration process would be:

1. Optionally pause all pipelines in the Deployer Concourse. eg:
   ```
   TARGET=... # master, prod, staging...
   ./bin/fly -t ${TARGET} pipelines -a | grep main | awk '{print $1}' | xargs -n1 -I {} ./bin/fly -t ${TARGET} pause-pipeline -p {}
   ``` 
2. Bring up new Bootstrap Concourse from `alphagov/paas-bootstrap`
    ```
    DEPLOY_ENV=... BRANCH=$(git rev-parse --abbrev-ref HEAD) make dev deployer-concourse bootstrap
    ```
1. Run the `create` pipeline on the Bootstrap Concourse.
   This will automatically: 
   	 * Create the new buckets and sync them.
   	 * Update the terraform resources (eg. vpc name)
   	 * update BOSH (new blobstore bucket)
   	 * migrate CF VMs for new blobstore, 
   	 * delete old concourse
   	 * recreate Concourse (takes approx 1hr)
   	 * attach the old disk to concourse and update bosh DB.
    
   Note: This failed for me once when updating `parser_z2/0` because it was unable to connect to Elasticsearch. The simplest, but time consuming, solution is to re-run the job.

1. Delete the cookies from your deployer. We found out that you cannot login after re-deployment having the old cookie set (???).

1. Run the  migration from paas-cf PR https://github.com/alphagov/paas-cf/pull/688:  Run `BRANCH=$(git rev-parse --abbrev-ref HEAD) make dev pipelines DEPLOY_ENV=....` (in https://github.com/alphagov/paas-cf/pull/688) to update all the pipelines in the new concourse (this will unpause them as part of the deployment).

Once this is finished:

1. Clean up manually:
	* Delete old S3 state bucket `<env>-state`. It's versioned, so you'll need to use the AWS console.
	* Delete old S3 blobstore bucket. `aws s3api delete-bucket --bucket ${DEPLOY_ENV}-bosh-blobstore`
1. Once all environments have been migrated (all), we can revert all the commits prefixed with `migrate:` in paas-bootstrap and paas-cf
1. ???
1. Profit


Who?
----

Anyone but @keymon, @dcarley or @alext
